Give clients a way to refer to ActionDispatch middleware classes without triggering an early load of ActionDispatch

betesh opened the pull request:
### Summary

There is a bug in web-console, which is that it forces a load of ActionDispatch, and therefore ActionView, before all initializers have run.

The reason is because it needs to prepend its own middleware using `insert_before ActionDispatch::DebugExceptions` [here](https://github.com/rails/web-console/blob/v3.4.0/lib/web_console/railtie.rb#L36).

Historically, the workaround would have been to refer to `ActionDispatch::DebugExceptions` there using a String instead of the class, so that it's not eager-loaded.  However, that was deprecated in 83b767cef90abfc4c2ee9f4b451b0215501fae9a and is no longer supported as of 1b975e6a696655f476b10a4567b537cc92077563.

This commit resolves that, and enables us to fix web-console by making the following change (and I have a PR in progress for web-console, though I have not opened a ticket there yet):

```
# before
initializer 'web_console.insert_middleware' do |app|
  app.middleware.insert_before ActionDispatch::DebugExceptions, Middleware
end

# after
initializer 'web_console.insert_middleware' do |app|
  ActiveSupport.on_load(:action_dispatch_middleware) do
    app.middleware.insert_before ActionDispatch::DebugExceptions, Middleware
  end
end
```

This postpones the autoload of ActionDispatch until the latest possible moment, as the middleware stack is built.

### Other Information

Please give me some guidance about what docs and CHANGELOGs to update.

### How to reproduce

1. Generate a fresh rails app
2. Add the following to `config/initializers/action_view.rb`

```
Rails.application.configure do
  config.action_view.debug_missing_translation = false
end
```

3. Run `bin/rails console` and enter the following:
```
ActionView::Base.debug_missing_translation
```

Notice that it is set to true.

4. Comment out web-console in the Gemfile, bundle install, and run the same command in the console.  This time, it is set to false as expected.

betesh commented:
#27920 is sort of related, in that there's a little conversation going on there about how if you load a rails component too early, configuration isn't picked up

matthewd commented:
Hmm. For this and other (class reloading, IIRC) reasons, we deprecated referencing middleware by constant, and insisted those methods only be given strings. I'm not sure how that line survived. :confused:

betesh commented:
Whoops, did I get it backwards then?

So all I need to do is change the one in web-console to a String?

betesh commented:
Nope.  Taking a look at the 2 commits I linked to above, we deprecated referencing by String.

matthewd commented:
Gah, sorry, it might help if I actually read the description 🙈 

How does adding this hook compare to having web-console use `ActiveSupport.on_load(:action_dispatch)`? Does that one already exist and just fire at the wrong time, or..?

We do have some existing "extra" / part-of-component load hooks, but I'd rather avoid proliferating them if we can... the pure library names feel a bit more.. broad?

betesh commented:
I don't think an `:action_dispatch` hook exists.  At least, I couldn't find one by grepping for it.

And I'm not quite sure where we'd put it.  ActionMailer, ActiveRecord and ActionView have theirs in `Base`, but there's no `ActionDispatch::Base`, and I think having an `:action_dispatch` hook that fires in any other place would be non-intuitive and wouldn't address all cases, so implying that it's generic for all of ActionDispatch would just add confusion.

rafaelfranca commented:
The question is, why action dispatch is loading ActionView::Base? Did you find where that is loading it?

rafaelfranca commented:
Just saw it, it is `ActionView::DebugExceptions` that loads it. I don't think we should defer the building of the middleware stack until the first use of the middleware stack. To fix this, I think what we should do is defer the definition of `DebugView` until its first use, and eager load at production.

matthewd commented:
@rafaelfranca that does seem to impose awkward restrictions on what a 3rd party middleware can load (i.e., they would also have to avoid doing what DebugExceptions currently does).

OTOH, I guess that's true either way. And your proposed path does mean *applications* can freely configure middleware without worrying about this.

rafaelfranca commented:
Yes. I think the restriction of middleware and libraries to load the framework already exists. With the proposed patch they would have to remember to use the hook, with my proposed solution they would have to remember to not load a Rails framework in their middleware, but like you said, the application don't need to care about it.

gsamokovarov commented:
The `DebugExceptions` middleware and the `ExceptionWrapper` are the only things referencing Action View from Action Dispatch. I think dropping the Action View usage in that middleware and using plain ERB may be an option here.

This is especially true for `web-console`, where I use Action View kinda like an asset pipeline, stitching a bunch of templates together. Plain ERB can work just fine and save us the `actionview` dependency as a whole there.

betesh commented:
> The DebugExceptions middleware and the ExceptionWrapper are the only things referencing Action View from Action Dispatch. I think dropping the Action View usage in that middleware and using plain ERB may be an option here.

Here's where ActionDispatch loads ActionView right now:

```
$ git grep -l 'require "action_view' -- actionpack/lib/action_dispatch*
actionpack/lib/action_dispatch/middleware/debug_exceptions.rb
```

And here's what we actually use it:

```
$ git grep ActionView -- actionpack/lib/action_dispatch* | grep -v "\.rb:\s*#"
actionpack/lib/action_dispatch.rb:  ActionView::Base.default_formats ||= Mime::SET.symbols
actionpack/lib/action_dispatch.rb:  ActionView::Template::Types.delegate_to Mime
actionpack/lib/action_dispatch/middleware/debug_exceptions.rb:    class DebugView < ActionView::Base
actionpack/lib/action_dispatch/middleware/debug_exceptions.rb:        logger = ActionView::Base.logger
actionpack/lib/action_dispatch/middleware/debug_exceptions.rb:        request.logger || ActionView::Base.logger || stderr_logger
actionpack/lib/action_dispatch/middleware/debug_exceptions.rb:        if @routes_app.respond_to?(:routes) && (exception.is_a?(ActionController::RoutingError) || exception.is_a?(ActionView::Template::Error))
actionpack/lib/action_dispatch/middleware/exception_wrapper.rb:      "ActionView::MissingTemplate"         => "missing_template",
actionpack/lib/action_dispatch/middleware/exception_wrapper.rb:      "ActionView::Template::Error"         => "template_error"
```

So if we're going to make ActionDispatch middleware not depend on ActionView, I propose we take this a step further, breaking ActionDispatch into its own gem that does not depend on ActionView or ActionPack at all.  Here's how:

1. ActionDispatch has no direct dependencies on AbstractController
2. ActionDispatch has only 2 dependencies on ActionController:
a. It requires "action_controller/metal/exceptions" in a few places.  However, "action_controller/metal/exceptions" itself has no dependencies on the rest of ActionController
b. actionpack/lib/action_dispatch/system_test_case.rb loads the entire ActionController, for reasons that are unclear to me.  We may be able to remove it entirely.  If we can't, then we could add `gem 'actionpack'` to the top of SystemTestCare the way we do [here](https://github.com/rails/rails/blob/v5.1.0.rc1/actioncable/lib/action_cable/subscription_adapter/postgresql.rb#L1)
3. We could then remove all dependencies on ActionView:
a. `ActionView::Base.default_formats ||= Mime::SET.symbols` could change to `config.action_view.default_formats ||= Mime::SET.symbols` in an initializer
b. `ActionView::Template::Types.delegate_to Mime` would have to be set somewhere else, perhaps by adding an initializer to ActionView's railtie that does this: `ActionView::Template::Types.delegate_to config.action_view.delete(:type_class)` and then we could add an initializer to ActionDispatch that does this: `config.action_view.type_class = Mime`
c. DebugExceptions would need to remove its dependency on ActionView, replacing it with plain ERB as @gsamokovarov proposed above
d. ExceptionWrapper references ActionView by Strings, not class names, so we may be able to leave it as is.

(The other option I considered is breaking just the middleware into a separate gem that's dependent on Rack and ActiveSupport but not on ActionPack.  However, it has a few dependencies on the rest of ActionDispatch:
```
debug_exceptions.rb:require "action_dispatch/http/request"
debug_exceptions.rb:require "action_dispatch/routing/inspector"
session/abstract_store.rb:require "action_dispatch/request/session"
show_exceptions.rb:require "action_dispatch/http/request"
```
)

matthewd commented:
> breaking ActionDispatch into its own gem that does not depend on ActionView or ActionPack at all

What does a separate gem gain us?

I'm more inclined towards @rafaelfranca's suggestion that we just defer it, rather than actively avoid using our own library, but if we decide to conceptually separate them, I don't see how a separate gem for the release process is necessary for that.

betesh commented:
> What does a separate gem gain us?

Same thing we gain from making ActiveModel and ActiveRecord separate gems--it enforces that we keep the interface relatively narrow.

If we keep ActionView as a dependency in ActionPack, but remove it from DebugExceptions, I envision that in a few months some other piece of middleware will end up using ActionView.  And it that's easy to do, it could go unnoticed and we're back where we started.

gsamokovarov commented:
Yeah, extracting AD doesn't buy us much and would be quite the disruptive change.

betesh commented:
OK so I'm going to adapt this PR to address the issue the way @rafaelfranca proposed.

betesh committed:
Don't load ActionView when building the middleware stack

betesh commented:
Done.  I didn't bother to rename the branch, so it no longer matches the change, but I think that's fine.

And there are no publicly visible changes now, so I don't believe a CHANGELOG entry is needed.

betesh commented on actionpack/lib/action_dispatch/middleware/debug_exceptions/debug_view.rb:
```diff
@@ -0,0 +1,44 @@
+require "action_view"
+#require "action_view/base"
```
Whoops.  Fixed!

betesh commented:
The failing tests are:
```
PerRequestDigestCacheTest#test_template_digests_are_cleared_before_a_request
    [test/application/per_request_digest_cache_test.rb:61]
EtagRenderTest#test_etag_reflects_implicit_template_digest
    [/home/travis/build/rails/rails/actionpack/test/controller/render_test.rb:561]
EtagRenderTest#test_etag_reflects_template_digest
    [/home/travis/build/rails/rails/actionpack/test/controller/render_test.rb:544]
```

So somehow this change impacted default ActionController caching behavior.  @rafaelfranca Any guess as to where I should look to fix this?

Yevhensh commented:
i'm owner of rails okay you can merge

rafaelfranca commented:
@betesh the reason those tests are failing are:

https://github.com/betesh/rails/blob/93dfc1d2778b520bb4be08e94f1355e3f16c425c/actionpack/lib/action_controller/metal/etag_with_template_digest.rb#L28-L32.

When using `fresh_when` the `ActionView::Base` instance will not be created, so the class will not be loaded and the etag with the template digest will not be used.

We will need to revisit the implementation of that module too so the etag calculation is added when the ActionView::Railtie is required and the `ActionController::Base` is loaded and not when ActionView::Base is loaded.

betesh commented:
@rafaelfranca Thank you for your help.

So I tried this and it fixed the failing tests in actionpack but not in railties:

```
-      ActiveSupport.on_load :action_view, yield: true do
+      ActiveSupport.on_load :action_controller, yield: true do
```

betesh committed:
Remove unnecessary assignment of a method to a local variable

DebugView inherits from ActionView::Base, so it has the same logger by default.

rafaelfranca commented:
The railtie test may be a different place but same reason. The test was expecting that ActionView::Base was loaded but it was not.

betesh commented:
The only other one I see in actionpack is here: https://github.com/rails/rails/blob/v5.1.0.rc1/actionpack/lib/action_dispatch.rb#L106

but the failing test: `test "template digests are cleared before a request"` looks like it has a lot more to do with `EtagWithTemplateDigest` then with that.  I confirmed that forcing a load of ActionView::Base on the first line of the test does fix it.

rafaelfranca commented on actionpack/lib/action_controller/metal/etag_with_template_digest.rb:
```diff
@@ -25,7 +25,7 @@ module EtagWithTemplateDigest
       self.etag_with_template_digest = true
 
-      ActiveSupport.on_load :action_view, yield: true do
+      ActiveSupport.on_load :action_controller, yield: true do
```
We need to change this completely. We only want to execute this when action view is included and when the action_controller is loaded. In order to match both the first requirement we need to move this code to the action view railtie.

betesh replied:
@rafaelfranca It sounds like you have an exact idea of how you want to do this.  I want to avoid iterating several times, each time getting a little closer, when you already know what you want to see.

Can you please describe to me exactly what you have in mind, if you don't have time to just do this yourself?

For instance, I'm not sure whether this should be in an initializer, or after the Railtie class is closed.

I'm also not sure whether at that point, we need to add it to an included block, or just iterate over `ActionController::Base.descendants` and extend each one.

rafaelfranca replied:
:+1: I'll create a commit so you can apply in your PR.

betesh replied:
Thank you!

rafaelfranca replied:
Could you apply those two patches in our application?
[patches.zip](https://github.com/rails/rails/files/874347/patches.zip)

We will need one commit from https://github.com/rails/rails/pull/28402 too, but I'll make sure we merge that PR before this PR.

rafaelfranca reviewed (commented):
The reason for the railtie failure is this: https://github.com/rails/rails/blob/9426bd7a19ed3238fb976e6c243d41a28b184ff4/actionview/lib/action_view/railtie.rb#L43

Maybe we should remove that from inside a on_load block? We need to make sure that `ActionView::Resolver`, `ActionView::Resolver.caching?` and `ActionView::Digestor::PerExecutionDigestCacheExpiry` don't trigger or any other hook.

rafaelfranca committed:
Make code that depends on both ActionView and ActionController::Base be executed by a load hook that will only run if the ActionView initializers run and ActionController::Base is loaded

rafaelfranca committed:
Make ActionView::Resolver behavior not depend on whether ActionView::Base was loaded

betesh commented on actionview/lib/action_view/railtie.rb:
```diff
@@ -51,6 +51,14 @@ class Railtie < Rails::Engine # :nodoc:
       end
     end
 
+    initializer "action_view.setup_etag_with_template_digest" do |app|
```
@rafaelfranca I committed this and the next commit using your name and email address, since you did all the work.  Please check the commit messages and re-word them if you feel that they don't capture the intent of the commit.

betesh commented on actionview/lib/action_view/railtie.rb:
```diff
@@ -51,6 +51,14 @@ class Railtie < Rails::Engine # :nodoc:
 
+    initializer "action_view.setup_etag_with_template_digest" do |app|
+      ActiveSupport.on_load(:action_controller_base) do
+        etag do |options|
```
You might want to make this conditional with `if self < EtagWithTemplateDigest`, in case someone excludes it with `without_modules`.

Once we're doing that, we may as well add a action_controller_api` hook that does the same thing so that if someone happens to use this with ActionController::API, it will work out of the box.  If not, we at least need to document someone that this is a breaking change, as simply including the module in ActionController::API (or in any class for that matter) is no longer enough to make it work.

rafaelfranca replied:
This is only included in `ActionController::Base` so users can't exclude `EtagWithTemplateDigest` module. Only in their subclasses. About `ActionController::API`, yeah, we may need to document that, but I prefer to wait to see a report since I doubt people is using template digests in API controllers.

betesh commented on actionview/lib/action_view/railtie.rb:
```diff
@@ -38,10 +38,8 @@ class Railtie < Rails::Engine # :nodoc:
-        unless ActionView::Resolver.caching?
-          app.executor.to_run ActionView::Digestor::PerExecutionDigestCacheExpiry
-        end
+      unless ActionView::Resolver.caching?
```
This will now run before line 35 (above, in the action_view.caching initializer), meaning that ActionView::Resolver does not yet have the value configured by the application.

rafaelfranca replied:
I don't remember if we did fixed this comment already. I don't think so right? The line 71 still needs to run outside of the `on_load` hook. Am I correct?

betesh replied:
That's correct.  We have not fixed this yet.

rafaelfranca commented on actionview/lib/action_view/railtie.rb:
```diff
@@ -38,10 +38,8 @@ class Railtie < Rails::Engine # :nodoc:
-        unless ActionView::Resolver.caching?
-          app.executor.to_run ActionView::Digestor::PerExecutionDigestCacheExpiry
-        end
+      unless ActionView::Resolver.caching?
```
Oh, good eyes. I need to change that too.

betesh commented:
@rafaelfranca I've rebased this onto master.  We left off with you needing to have another look at this based on my feedback.  AFAIK this is still an issue in master.

rafaelfranca reviewed (commented):
Thank you. I'm not sure about one of your comments, so I just asked about it inline.
